File: src/media-service/src/decorators/grpc_server_decorator.rs
use std::sync::Arc;
use std::sync::atomic::{AtomicU64, Ordering};
use crate::repositories::post_repo::PostRepository;
use crate::services::grpc_server::post_server::post_service_server::PostService;
use crate::services::grpc_server::post_server::{BatchPostsRequest, PaginatedPostsResponse, PostRequest, PostResponse, PostsRequest, PostsResponse};
use crate::services::grpc_server::GrpcPostServer;
use opentelemetry::metrics::{Counter, Histogram};
use opentelemetry::{global, KeyValue};
use tonic::{Code, Request, Response, Status};
use tracing::Span;
use tracing_opentelemetry::OpenTelemetrySpanExt;
use crate::utils::constants::OTEL_SERVICE_NAME;
use crate::utils::helpers::get_container_id;

pub struct ObservablePostServer<P: PostService + 'static> {
    inner: P,
    request_count: Counter<u64>,
    request_latency: Histogram<f64>,
    active_requests: Arc<AtomicU64>,
    attributes: Vec<KeyValue>,
}

impl<P: PostService + 'static> ObservablePostServer<P> {
    pub fn new(inner: P) -> Self {
        let boundaries = vec![
            0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0,
        ];

        let meter_provider = global::meter(OTEL_SERVICE_NAME);
        let request_count = meter_provider
            .u64_counter("grpc_server_requests_total")
            .with_description("Total number of gRPC requests")
            .build();

        let request_latency = meter_provider
            .f64_histogram("grpc_server_request_duration_seconds")
            .with_description("Request processing duration")
            .with_boundaries(boundaries)
            .build();
        
        let host_name = get_container_id().unwrap_or(String::from("0.0.0.0"));
        let attributes = vec![
            KeyValue::new("service", OTEL_SERVICE_NAME),
            KeyValue::new("instance", host_name),
            KeyValue::new("env", std::env::var("APP_ENV").unwrap_or(String::from("development")))
        ];
        let active_requests = Arc::new(AtomicU64::new(0));
        let active_requests_clone = active_requests.clone();

        let attributes_clone = attributes.clone();
        meter_provider
            .u64_observable_gauge("grpc_server_active_requests")
            .with_description("Active gRPC requests")
            .with_callback(move |observer| {
                let value = active_requests_clone.load(Ordering::Relaxed);
                observer.observe(value, &attributes_clone);
            })
            .build();

        Self {
            inner,
            request_count,
            request_latency,
            active_requests,
            attributes
        }
    }

    async fn track_method<T, F>(
        &self,
        method_name: &str,
        operation: F,
    ) -> Result<Response<T>, Status>
    where
        F: Future<Output = Result<Response<T>, Status>>,
    {
        let start_time = tokio::time::Instant::now();
        self.active_requests.fetch_add(1, Ordering::Relaxed);
        
        let result = operation.await;
        let status = if result.is_ok() { "success" } else { "error" };

        let mut attributes = vec![
            KeyValue::new("method", method_name.to_string()),
        ];
        
        attributes.extend_from_slice(&self.attributes);
        self.request_latency
            .record(start_time.elapsed().as_secs_f64(), &attributes);

        attributes.push(KeyValue::new("status", status));
        self.request_count.add(1, &attributes);
        self.active_requests.fetch_sub(1, Ordering::Relaxed);
        
        let span = Span::current();
        if let Err(ref err) = result {
            let code = err.code();
            span.record("rpc.grpc.status_code", code.to_string());

            match code {
                Code::Unknown
                | Code::DeadlineExceeded
                | Code::Unimplemented
                | Code::Internal
                | Code::Unavailable
                | Code::DataLoss => {
                    span.record("error.type", "grpc_server_error")
                        .set_status(opentelemetry::trace::Status::error(err.to_string()));
                }
                _ => {}
            }

            return result;
        }

        span.set_status(opentelemetry::trace::Status::Ok);
        
        result
    }
}

#[tonic::async_trait]
impl<P> PostService for ObservablePostServer<GrpcPostServer<P>>
where
    P: PostRepository + 'static,
{
    async fn get_post_by_id(&self, request: Request<PostRequest>) -> Result<Response<PostResponse>, Status> {
        self.track_method(
            "GetPostById",
            self.inner.get_post_by_id(request),
        )
            .await
    }

    async fn get_paginated_posts(&self, request: Request<PostsRequest>) -> Result<Response<PaginatedPostsResponse>, Status> {
        self.track_method(
            "GetPaginatedPosts",
            self.inner.get_paginated_posts(request),
        )
            .await
    }

    async fn get_batch_posts(&self, request: Request<BatchPostsRequest>) -> Result<Response<PostsResponse>, Status> {
        self.track_method(
            "GetBatchPosts",
            self.inner.get_batch_posts(request),
        )
            .await
    }
}

File: src/social-graph/internal/decorators/grpc_server_decorator.go
package decorators

import (
	"context"
	"fmt"
	"github.com/mqsrr/zylo/social-graph/internal/config"
	"github.com/mqsrr/zylo/social-graph/internal/protos/github.com/mqsrr/zylo/social-graph/proto"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
	"sync/atomic"
	"time"
)

type ObservableRelationshipServiceServer struct {
	proto.UnimplementedRelationshipServiceServer
	inner          proto.RelationshipServiceServer
	requestCounter metric.Int64Counter
	requestLatency metric.Float64Histogram
	activeRequests int64
	attributes     attribute.Set
}

func NewObservableRelationshipServer(inner proto.RelationshipServiceServer, meterProvider metric.MeterProvider) (*ObservableRelationshipServiceServer, error) {
	meter := meterProvider.Meter(config.DefaultConfig.ServiceName)
	requestCounter, err := meter.Int64Counter("grpc_server_requests_total", metric.WithDescription("Total number of gRPC requests"))
	if err != nil {
		return nil, err
	}

	requestLatency, err := meter.Float64Histogram("grpc_server_request_duration_seconds",
		metric.WithDescription("Request processing duration"),
		metric.WithExplicitBucketBoundaries(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0))

	if err != nil {
		return nil, err
	}

	containerId, err := config.GetContainerID()
	if err != nil {
		containerId = "0.0.0.0"
	}

	attributes := attribute.NewSet(attribute.String("service", config.DefaultConfig.ServiceName),
		attribute.String("instance", fmt.Sprintf("%s:%s", containerId, config.DefaultConfig.Port)),
		attribute.String("env", config.DefaultConfig.Environment))

	var activeRequests int64
	_, err = meter.Int64ObservableGauge("grpc_server_active_requests",
		metric.WithDescription("Active gRPC requests"),
		metric.WithInt64Callback(func(ctx context.Context, o metric.Int64Observer) error {
			o.Observe(activeRequests, metric.WithAttributeSet(attributes))
			return nil
		}))

	if err != nil {
		return nil, err
	}

	return &ObservableRelationshipServiceServer{
		inner:          inner,
		requestCounter: requestCounter,
		requestLatency: requestLatency,
		activeRequests: activeRequests,
		attributes:     attributes,
	}, nil
}

func (s *ObservableRelationshipServiceServer) GetUserRelationships(ctx context.Context, req *proto.RelationshipRequest) (*proto.RelationshipResponse, error) {
	atomic.AddInt64(&s.activeRequests, 1)
	startTime := time.Now()

	ok, err := s.inner.GetUserRelationships(ctx, req)
	duration := time.Since(startTime).Seconds()

	status := "success"
	if err != nil {
		status = "error"
	}

	s.requestCounter.Add(ctx, 1,
		metric.WithAttributes(
			attribute.String("method", "GetUserRelationships"),
			attribute.String("status", status)),
		metric.WithAttributeSet(s.attributes))

	s.requestLatency.Record(ctx, duration, metric.WithAttributes(
		attribute.String("method", "GetUserRelationships"),
		attribute.String("status", status)),
		metric.WithAttributeSet(s.attributes))

	atomic.AddInt64(&s.activeRequests, -1)
	return ok, err
}

func (s *ObservableRelationshipServiceServer) GetBatchRelationships(ctx context.Context, req *proto.BatchRelationshipRequest) (*proto.BatchRelationshipResponse, error) {
	atomic.AddInt64(&s.activeRequests, 1)

	startTime := time.Now()
	ok, err := s.inner.GetBatchRelationships(ctx, req)
	duration := time.Since(startTime).Seconds()

	status := "success"
	if err != nil {
		status = "error"
	}

	s.requestCounter.Add(ctx, 1,
		metric.WithAttributes(
			attribute.String("method", "GetBatchRelationships"),
			attribute.String("status", status)),
		metric.WithAttributeSet(s.attributes))

	s.requestLatency.Record(ctx, duration, metric.WithAttributes(
		attribute.String("method", "GetBatchRelationships"),
		attribute.String("status", status)),
		metric.WithAttributeSet(s.attributes))

	atomic.AddInt64(&s.activeRequests, -1)
	return ok, err
}
